src/vm/lib/metadata/agent.js

```diff
@@ -574,299 +763,243 @@ function (zonename, callback) {
-         *     '<zoneroot>/foo/bar',
-         *     '<zoneroot>/foo'
-         *  ]
-         *
-         * and then attempt to ensure each component exists.
-         */
-        d = sockdir;
-        while (d.length > zoneroot.length) {
-            dirs.push(d);
-            d = path.dirname(d);
-        }
+    var zlog = self.zlog[zopts.zone];
 
-        assert.ok(dirs.length > 0, 'should have at least one dir');
+    if (!zlog) {
+        // if there's no zone-specific logger, use the global one
+        zlog = self.log;
+    }
 
-        vasync.forEachPipeline({
-            inputs: dirs,
-            func: _ensureDir
-        }, callback);
+    if (!self.zones[zopts.zone]) {
+        zlog.info({zonename: zopts.zone},
+            'zone no longer exists, not creating zsock');
+        callback();
+        return;
     }
 
-    function _attemptCreate(callback) {
-        _ensureSockpathDirsExist(function _ensureSockpathDirsExistCb(error) {
-            if (error) {
-                callback(error);
-                return;
-            }
-            zsock.createZoneSocket(zopts, callback);
-        });
+    // Ignore zones we've already got a connection for and then immediately
+    // create an entry if we don't. To act as a mutex.
+    if (self.zoneConnections[zopts.zone]) {
+        zlog.trace({zonename: zopts.zone}, 'already have zoneConnections[] for '
+            + 'zone -- not replacing.');
+        callback();
+        return;
     }
+    self.zoneConnections[zopts.zone] = {};
 
     self.addDebug(zopts.zone, 'last_zsock_create_attempt');
-    _attemptCreate(function (error, fd) {
-        var server;
-
+    _ensureSockpathDirsExist(zopts, zlog, function _ensureDirsExistCb(error) {
         if (error) {
-            zoneExists(zopts.zone, function _zoneExists(_, exists) {
-                if (!exists) {
-                    zlog.warn({err: error}, 'error creating socket and zone no '
-                        + 'longer exists, not retrying');
-                    return;
-                }
-                // If we get errors trying to create the zone socket, setup a
-                // retry loop and return.
-                zlog.error({err: error}, 'createZoneSocket error, %s seconds '
-                    + 'before next attempt', waitSecs);
-                _retryCreateZoneSocketLater();
-                return;
-            });
+            self.addDebug(zopts.zone, 'last_zsock_create_failure', error);
+
+            zlog.warn({zonename: zopts.zone, err: error},
+                'failed to create sockpath directory');
+
+            // We were unable to create the directory but we have not yet
+            // created a real self.zoneConnections entry so we can just delete
+            // the placeholder and call callback. If the VM still exists and
+            // is running, we'll try again when we next poll w/
+            // _checkNewZones().
+            delete self.zoneConnections[zopts.zone];
+            callback();
```

Comment:
Is it working?
On Apr 12, 2016 1:11 AM, "Josh Wilsdon" <notifications@github.com> wrote:

> In src/vm/lib/metadata/agent.js
> <https://github.com/joyent/smartos-live/pull/594#discussion_r59324962>:
>
> > -                    + 'before next attempt', waitSecs);
> > -                _retryCreateZoneSocketLater();
> > -                return;
> > -            });
> > +            self.addDebug(zopts.zone, 'last_zsock_create_failure', error);
> > +
> > +            zlog.warn({zonename: zopts.zone, err: error},
> > +                'failed to create sockpath directory');
> > +
> > +            // We were unable to create the directory but we have not yet
> > +            // created a real self.zoneConnections entry so we can just delete
> > +            // the placeholder and call callback. If the VM still exists and
> > +            // is running, we'll try again when we next poll w/
> > +            // _checkNewZones().
> > +            delete self.zoneConnections[zopts.zone];
> > +            callback();
>
> The only place this gets called is startZoneSocketServer() none of whose
> callbacks were doing anything with the error. However, with the changes you
> suggested wrt updating the "created" counter only when successful, this
> will change. So I'll pass the error back for that to work.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/joyent/smartos-live/pull/594/files/9a0569cd80d068189e7af8b9a9de73c75d9fc4a5#r59324962>
>